Marlin/MarlinSerial.cpp

```diff
@@ -310,3 +315,98 @@ MarlinSerial customizedSerial;
 #if defined(USBCON) && ENABLED(BLUETOOTH)
   HardwareSerial bluetoothSerial;
 #endif
+
+#if ENABLED(EMERGENCY_PARSER)
+
+  // Currently looking for: M108, M112, M410
+  // If you alter the parser please don't forget to update the capabilities in Conditionals.h
+
+  void emergency_parser(unsigned char c) {
+
+    enum e_parser_state {
+      state_RESET,
+      state_N,
+      state_M,
+      state_M1,
+      state_M10,
+      state_M108,
+      state_M11,
+      state_M112,
+      state_M4,
+      state_M41,
+      state_M410,
+      state_IGNORE // to '\n'
+    };
+
+    static e_parser_state state = state_RESET;
+
+    switch (state) {
+      case state_RESET:
+        switch (c) {
+          case ' ': break;
+          case 'N': state = state_N;      break;
+          case 'M': state = state_M;      break;
+          default: state = state_IGNORE;
+        }
+        break;
+
+      case state_N:
+        if (!NUMERIC_SIGNED(c) && c != ' ')
+          state = (c == 'M') ? state_M : state_RESET;
+        break;
```

Comment:
I'm quite unlucky `with state_N`. Now we have 6 decisions and a assignment. In my original version there have been always 2 decisions and one assignment. The beauty got lost.